In [100]:
import os
os.chdir(os.getcwd() + "\..")

In [101]:
from src import damage

## Кирара

### Базовые статы + таланты

In [114]:
kirara_base_hp = 12180
kirara_base_atk = 223
kirara_hp_ascention = 0.24

def kirara_shield_absorbtion(hp):
    return 0.17 * hp + 1956

def kirara_max_shield_absorbtion(hp):
    return 0.272 * hp + 3133

def kirara_reinforced_packaging(hp, stacks=3):
    return kirara_shield_absorbtion(hp) * 0.2 * stacks

def kirara_min_reinf_stacked(hp, stacks=3):
    return kirara_shield_absorbtion(hp) + kirara_reinforced_packaging(hp, stacks=stacks)

kirara_ult_mv = 9.694 + 0.606 * 4

### Билд

Допустим, дабл ХП + 40% из подстатов. В шапке крит, чтоб фав факинг прокал. Допустим она на дендро сете.

In [115]:
artifacts_hp_percent = 0.466 * 2 + 0.4
artifacts_flat_hp = 4000
# Реализуемо ок? ок
artifacts_atk_percent = 0.2
artifacts_flat_atk = 400
favge_base = 454

### Сколько же танчит ее щит?

In [117]:
kirara_total_hp = damage.total_hp(kirara_base_hp, artifacts_hp_percent + kirara_hp_ascention, artifacts_flat_hp)
kirara_total_hp

35326.96

Как стакается А4:

Геовишап

Удар1: 4164

Удар2: 5205

Удар3: 6505

In [97]:
print(damage.shield_hp(kirara_shield_absorbtion(16955), 1))
print(damage.shield_hp(kirara_min_reinf_stacked(16955, stacks=2), 1))
print(damage.shield_hp(kirara_min_reinf_stacked(16955, stacks=3), 1))

4838.35
6773.6900000000005
7741.360000000001


In [118]:
# Все сходится: урон 3го удара геовишапа при слабом щите
6505 - 4838

1667

In [119]:
print("Урон со 3го удара (после 2го) при двух стаках:", 6505 + 5205 - 6773)
print("Урон со 3го удара (после 2го) при трех стаках:", 6505 + 5205 - 7741)

Урон со 3го удара (после 2го) при двух стаках: 4937
Урон со 3го удара (после 2го) при трех стаках: 3969


Базовое ХП щита и максимальное

In [120]:
damage.shield_hp(kirara_shield_absorbtion(kirara_total_hp), 1)

7961.5832

In [121]:
damage.shield_hp(kirara_max_shield_absorbtion(kirara_total_hp), 1)

12741.93312

Вместе со стаками А1 начальное значение должно быть больше

Дополнительные щиты (ХП) от ее А1. За одну ешку смогут настакать почти до максимума.

In [122]:
damage.shield_hp(kirara_reinforced_packaging(kirara_total_hp), 1)

4776.94992

In [123]:
damage.shield_hp(kirara_min_reinf_stacked(kirara_total_hp), 1)

12738.53312

Но измерения лишь ХП щита не дают полной картины о его прочности. Нужно также учитывать его кд. Берем КД 8, но максимальный составляет 12. Мы не хотим, чтобы он составлял 12.

In [124]:
print(damage.shield_hp(kirara_min_reinf_stacked(kirara_total_hp), 1)/9)
print(damage.shield_hp(kirara_max_shield_absorbtion(kirara_total_hp), 1)/8)

1415.392568888889
1592.74164


Что если урон - дендро (что вполне реально в блум и барджион командах)

In [125]:
print(damage.shield_hp(kirara_min_reinf_stacked(kirara_total_hp), 1, damage_absorbtion=2.5)/9)
print(damage.shield_hp(kirara_max_shield_absorbtion(kirara_total_hp), 1, damage_absorbtion=2.5)/8)

3538.481422222222
3981.8541


Под Милилитом (не реалистично)

In [126]:
print(damage.shield_hp(kirara_min_reinf_stacked(kirara_total_hp), 1, shield_strength=1.3)/12)
print(damage.shield_hp(kirara_max_shield_absorbtion(kirara_total_hp), 1, shield_strength=1.3)/12)

1380.0077546666669
1380.376088


### Сколько же наносит ее ульта?

In [127]:
kirara_total_atk = damage.total_atk(kirara_base_atk + favge_base, artifacts_atk_percent, artifacts_flat_atk)
kirara_a4_ult_buff = kirara_total_hp / 1000 * 0.003
effective_talent = damage.effective_talent(kirara_total_atk, cd=1, cr=0.5, crit_ratio=0.5, dmg_bonus=1 + 0.15 + kirara_a4_ult_buff)
talent_dmg = damage.talent_dmg(effective_talent, kirara_ult_mv)
outgoing_dmg = damage.outgoing_dmg_multiplicative(talent_dmg, character_level=90, enemy_level=100, enemy_resistance=-0.2) # Срез резиста от дендро сета
outgoing_dmg

26643.12906304656

In [128]:
kirara_a4_ult_buff

0.10598088

10% BAAAATCHEEEST

## Дед

### Базовые статы + таланты

In [129]:
ded_base_hp = 14695
ded_base_atk = 251

def ded_shield_absorbtion(hp):
    return 0.218 * hp + 2506


Билд тот же самый, что и у Кирары (и тоже дендро сет, потому что дед может прокать дендро сет)

### Что по щиту? - самый прочный в игре

In [130]:
ded_total_hp = damage.total_hp(ded_base_hp, artifacts_hp_percent, artifacts_flat_hp)
ded_total_hp

38268.74

In [131]:
damage.shield_hp(ded_shield_absorbtion(ded_total_hp), 1, shield_strength=1 + 0.05*5, damage_absorbtion=1.5)

20341.097475

ХП щита за секунду

In [132]:
damage.shield_hp(ded_shield_absorbtion(ded_total_hp), 1, shield_strength=1 + 0.05*5, damage_absorbtion=1.5) / 12

1695.09145625

## И что в итоге:

Кирара:

- Макс проч. 1592 ХП/сек, на деле 1415 ХП/сек (3500 ХП/сек против дендро) по-хорошему: 9 секунд, иначе 12 сек

- 100% щита

- Хороший нюк, который скорее всего будешь использовать

- Энергонезависимая + ДЕДНРО батарейка

- Режет дендро резист на 30%

- Она может накладывать деднро (?)

- Деднро: резонанс + не выключает Нилу

- Миллилит почти не работает

- Кошкодевочка


Дед:

- Макс проч. 1695 ХП/сек, по-хорошему на 12 сек

- 100% щита

- Нюк, который не будещь использовать (петрификация бесполезная кста)

- Энергонезависимый + гео

- Режет дендро на 30% и все 20%

- Может накладывать ничего

- Гео

- Может носить Миллилит

- Какой-то старый дед

А просто что если: Кирара сможет прокать Милилит из вне поля?

In [134]:
print(damage.shield_hp(kirara_min_reinf_stacked(kirara_total_hp), 1, shield_strength=1.6)/9)
print(damage.shield_hp(kirara_max_shield_absorbtion(kirara_total_hp), 1, shield_strength=1.6)/8)

2264.6281102222224
2548.386624


In [135]:
damage.shield_hp(ded_shield_absorbtion(ded_total_hp), 1, shield_strength=1 + 0.05*5 + 0.6, damage_absorbtion=1.5) / 12

2508.7353552500003